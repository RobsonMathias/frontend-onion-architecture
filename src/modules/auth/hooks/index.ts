export * from './use-auth';
export * from './use-login';
export * from './use-sign-up';
