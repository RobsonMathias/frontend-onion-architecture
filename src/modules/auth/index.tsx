export * from './login';
export * from './sign-up';
export * from './external-auth';
export * from './logout';
export * from './hooks';
