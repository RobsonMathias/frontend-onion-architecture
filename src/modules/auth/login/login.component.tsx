import React, { useEffect } from 'react';

import { LoginPayload } from '@types';
import { FieldErrors, useForm } from 'react-hook-form';
import { Box, Button, Grid, TextField } from '@mui/material';

export type LoginComponentProps = {
  onSubmit: (payload: LoginPayload) => void;
  loading?: boolean;
  action?: React.ReactNode;
};

export function LoginComponent(props: LoginComponentProps) {
  const { onSubmit, loading, action } = props;

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isSubmitted }
  } = useForm({ mode: 'all' });

  function hasError(error?: FieldErrors): string {
    const { message } = error || { message: '' };
    return isSubmitted ? message : '';
  }

  function submit(data: LoginPayload) {
    onSubmit(data);
  }

  useEffect(() => {
    return () => {
      reset();
    };
  }, []);

  return (
    <form onSubmit={handleSubmit(submit)}>
      <Box mb={2}>
        <TextField
          variant="filled"
          fullWidth={true}
          type="email"
          inputProps={{ 'data-testid': 'email' }}
          label="E-mail"
          error={!!hasError(errors.email)}
          helperText={hasError(errors.email)}
          placeholder="email@exemplo.com"
          {...register('email', {
            required: 'E-mail é obrigatório.'
          })}
        />
      </Box>
      <Box mb={2}>
        <TextField
          type="password"
          variant="filled"
          inputProps={{ 'data-testid': 'password' }}
          fullWidth={true}
          autoComplete="off"
          label="Senha"
          placeholder="Digite sua senha"
          error={!!hasError(errors.password)}
          helperText={hasError(errors.password)}
          {...register('password', {
            minLength: {
              value: 9,
              message: 'Senha precisa ter no mínimo 9 caracteres.'
            },
            required: 'Senha é obrigatória.'
          })}
        />
      </Box>
      <Grid container spacing={2} justifyContent="end">
        {action && <Grid item>{action}</Grid>}
        <Grid item>
          <Button
            disabled={loading}
            data-testid="submit"
            type="submit"
            variant="contained"
            color="secondary"
          >
            Acessar
          </Button>
        </Grid>
      </Grid>
    </form>
  );
}
