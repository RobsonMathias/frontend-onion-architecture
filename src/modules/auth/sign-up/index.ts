export * from './sign-up.container';
