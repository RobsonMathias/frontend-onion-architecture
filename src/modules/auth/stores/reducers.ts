import { combineReducers } from '@reduxjs/toolkit';
import user from './user.store';

export const rootReducer = combineReducers({
  user
});
