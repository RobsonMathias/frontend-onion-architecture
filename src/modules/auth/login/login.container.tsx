import React from 'react';

import { LoginComponent } from './login.component';
import { useLogin } from '../hooks';
import { LoginPayload } from '@types';
import { STORE_STATUS_LOADING } from 'constants/store';

export interface LoginContainerProps {
  action?: React.ReactNode;
}

export function LoginContainer(props: LoginContainerProps) {
  const { action } = props;
  const { login, auth } = useLogin();

  function handleSubmit(form: LoginPayload) {
    login(form.email, form.password);
  }

  return (
    <>
      <LoginComponent
        action={action}
        onSubmit={handleSubmit}
        loading={auth.user.status === STORE_STATUS_LOADING}
      />
    </>
  );
}
