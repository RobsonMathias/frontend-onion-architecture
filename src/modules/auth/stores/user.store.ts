import { createAsyncThunk } from '@reduxjs/toolkit';
import { STORE_AUTH_FETCH_USER, STORE_AUTH, STORE_STATUS_INITIALIZED } from 'constants/store';
import { Auth, ErrorStore } from '@types';
import { createGenericSlice, GenericState, createReducers, ExtraReducers } from 'stores/generics';

import { Payload } from 'stores';

export type AuthUserState = Auth;

type State = GenericState<AuthUserState>;

export const initialState: State = {
  data: {
    displayName: '',
    email: '',
    id: '',
    token: ''
  },
  status: STORE_STATUS_INITIALIZED,
  error: null
};

export const fetchAuth = createAsyncThunk<AuthUserState, Promise<Auth>, Payload>(
  STORE_AUTH_FETCH_USER,
  async (promise, { rejectWithValue }) => {
    try {
      const response = await promise;
      return response;
    } catch (e: unknown) {
      const { code, message } = e as ErrorStore;
      return rejectWithValue({ code, message });
    }
  }
);

const extraReducers: ExtraReducers<AuthUserState> = createReducers<AuthUserState, Auth, Payload>(
  fetchAuth,
  initialState.data
);

const slice = createGenericSlice({
  name: STORE_AUTH,
  initialState,
  reducers: {},
  extraReducers
});

const { actions, reducer } = slice;

const { clear } = actions;

export { clear };

export default reducer;
