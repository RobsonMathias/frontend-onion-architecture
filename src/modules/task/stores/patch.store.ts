import { createAsyncThunk } from '@reduxjs/toolkit';
import { STORE_TASK, STORE_TASK_PATCH, STORE_STATUS_INITIALIZED } from 'constants/store';
import { ErrorStore, Task, TaskResponse, TaskStatus } from '@types';
import { createGenericSlice, createReducers, ExtraReducers, GenericState } from 'stores/generics';

import { Payload } from 'stores';

export type PatchTaskState = Task;

type State = GenericState<PatchTaskState>;

export const initialState: State = {
  data: {
    description: '',
    id: '',
    status: TaskStatus.PROGRESS
  },
  status: STORE_STATUS_INITIALIZED,
  error: null
};

export const patchTask = createAsyncThunk<PatchTaskState, Promise<TaskResponse>, Payload>(
  STORE_TASK_PATCH,
  async (promise, { rejectWithValue }) => {
    try {
      const response = await promise;
      return response;
    } catch (e: unknown) {
      const { code, message } = e as ErrorStore;
      return rejectWithValue({ code, message });
    }
  }
);

const extraReducers: ExtraReducers<PatchTaskState> = createReducers<
  PatchTaskState,
  TaskResponse,
  Payload
>(patchTask, initialState.data);

const slice = createGenericSlice({
  name: STORE_TASK,
  initialState,
  reducers: {},
  extraReducers
});

const { actions, reducer } = slice;

const { clear } = actions;

export { clear };

export default reducer;
