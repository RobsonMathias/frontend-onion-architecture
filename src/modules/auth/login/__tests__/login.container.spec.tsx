/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect';

import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { STORE_AUTH_FETCH_USER_MOCK_FINISHED as mockAuthUserStore } from '@fixtures/stores/auth';
import { mockAuthStore } from '@fixtures/stores/auth';

function Mock({ onSubmit }: LoginComponentProps) {
  return (
    <div
      data-testid="mock"
      onClick={() =>
        onSubmit({
          email: 'email',
          password: 'password'
        })
      }
    />
  );
}

const mockLogin = jest.fn();
jest.mock('../../hooks', () => {
  return {
    useLogin: jest.fn(() => ({
      login: jest.fn((...args) => mockLogin(...args)),
      auth: {
        ...mockAuthStore.auth,
        user: mockAuthUserStore
      }
    }))
  };
});
jest.mock(__dirname + '/../login.component', () => ({
  LoginComponent: Mock
}));

import { LoginContainer } from '../login.container';
import { LoginComponentProps } from '../login.component';

describe('LoginContainer', () => {
  it('should render successfully', () => {
    const { container } = render(<LoginContainer />);
    expect(container).toMatchSnapshot();
  });

  it('should call handleSubmit successfully', async () => {
    const { getByTestId } = render(<LoginContainer />);

    await act(async () => {
      fireEvent.click(getByTestId('mock'));
    });

    expect(mockLogin).toHaveBeenCalledWith('email', 'password');
  });
});
