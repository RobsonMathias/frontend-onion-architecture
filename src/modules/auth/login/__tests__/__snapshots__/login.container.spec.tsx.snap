// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginContainer should render successfully 1`] = `
<div>
  <div
    data-testid="mock"
  />
</div>
`;
